tiny_cnn/network.h

```diff
@@ -732,17 +745,15 @@ class network
         return std::abs(delta_by_bprop - delta_by_numerical) <= eps;
     }
 
+    // convenience wrapper for the function below
     template <typename E>
-    void bprop(const tensor_t& out, const tensor_t& t, int idx, const tensor_t* t_cost) {
-        tensor_t delta = gradient<E>(out, t);
-
-        if (t_cost) {
-            for_i(delta.size(), [&](int i) {
-                for (size_t j = 0; j < delta[i].size(); j++)
-                    delta[i][j] *= (*t_cost)[i][j];
-            });
-        }
+    void bprop(const std::vector<vec_t>& out, const std::vector<vec_t>& t, int idx, const std::vector<vec_t>& t_cost) {
+        bprop<E>(std::vector<tensor_t>{out}, std::vector<tensor_t>{t}, idx, std::vector<tensor_t>{t_cost});
+    }
 
+    template <typename E>
+    void bprop(const std::vector<tensor_t>& out, const std::vector<tensor_t>& t, int idx, const std::vector<tensor_t>& t_cost) {
+        tensor_t delta = gradient<E>(out, t, t_cost);
         net_.backward(delta, idx);
```

Comment:
This too may require some further changes when implementing Batch Normalization.